In [169]:
import pandas as pd
import numpy as np

Loaded the Dataset

In [170]:
df =pd.read_csv(r"D:\Mega Project\AI-POWERED-ITSM\data\raw\incident_event_log.csv")

Inspecting the values

In [171]:
df.head(5)

,number,incident_state,active,reassignment_count,reopen_count,sys_mod_count,made_sla,caller_id,opened_by,opened_at,...,u_priority_confirmation,notify,problem_id,rfc,vendor,caused_by,closed_code,resolved_by,resolved_at,closed_at
0,INC0000045,New,True,0,0,0,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
1,INC0000045,Resolved,True,0,0,2,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
2,INC0000045,Resolved,True,0,0,3,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
3,INC0000045,Closed,False,0,0,4,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
4,INC0000047,New,True,0,0,0,True,Caller 2403,Opened by 397,29/2/2016 04:40,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 81,1/3/2016 09:52,6/3/2016 10:00


Size of the dataset

In [172]:
df.shape

(141712, 36)

Column names

In [173]:
df.columns

Index(['number', 'incident_state', 'active', 'reassignment_count',
       'reopen_count', 'sys_mod_count', 'made_sla', 'caller_id', 'opened_by',
       'opened_at', 'sys_created_by', 'sys_created_at', 'sys_updated_by',
       'sys_updated_at', 'contact_type', 'location', 'category', 'subcategory',
       'u_symptom', 'cmdb_ci', 'impact', 'urgency', 'priority',
       'assignment_group', 'assigned_to', 'knowledge',
       'u_priority_confirmation', 'notify', 'problem_id', 'rfc', 'vendor',
       'caused_by', 'closed_code', 'resolved_by', 'resolved_at', 'closed_at'],
      dtype='str')

In [174]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 141712 entries, 0 to 141711
Data columns (total 36 columns):
 #   Column                   Non-Null Count   Dtype
---  ------                   --------------   -----
 0   number                   141712 non-null  str  
 1   incident_state           141712 non-null  str  
 2   active                   141712 non-null  bool 
 3   reassignment_count       141712 non-null  int64
 4   reopen_count             141712 non-null  int64
 5   sys_mod_count            141712 non-null  int64
 6   made_sla                 141712 non-null  bool 
 7   caller_id                141712 non-null  str  
 8   opened_by                141712 non-null  str  
 9   opened_at                141712 non-null  str  
 10  sys_created_by           141712 non-null  str  
 11  sys_created_at           141712 non-null  str  
 12  sys_updated_by           141712 non-null  str  
 13  sys_updated_at           141712 non-null  str  
 14  contact_type             141712 non-null  str  

Checking the data types

In [175]:
df.dtypes

number                       str
incident_state               str
active                      bool
reassignment_count         int64
reopen_count               int64
sys_mod_count              int64
made_sla                    bool
caller_id                    str
opened_by                    str
opened_at                    str
sys_created_by               str
sys_created_at               str
sys_updated_by               str
sys_updated_at               str
contact_type                 str
location                     str
category                     str
subcategory                  str
u_symptom                    str
cmdb_ci                      str
impact                       str
urgency                      str
priority                     str
assignment_group             str
assigned_to                  str
knowledge                   bool
u_priority_confirmation     bool
notify                       str
problem_id                   str
rfc                          str
vendor    

In [176]:
df.describe()

,reassignment_count,reopen_count,sys_mod_count
count,141712.000000,141712.000000,141712.000000
mean,1.104197,0.021918,5.080946
std,1.734673,0.207302,7.680652
min,0.000000,0.000000,0.000000
25%,0.000000,0.000000,1.000000
50%,1.000000,0.000000,3.000000
75%,1.000000,0.000000,6.000000
max,27.000000,8.000000,129.000000


Checking column Variables which are not numerical

In [177]:
cols=df.select_dtypes(include='object').columns
print(cols)

Index(['number', 'incident_state', 'caller_id', 'opened_by', 'opened_at',
       'sys_created_by', 'sys_created_at', 'sys_updated_by', 'sys_updated_at',
       'contact_type', 'location', 'category', 'subcategory', 'u_symptom',
       'cmdb_ci', 'impact', 'urgency', 'priority', 'assignment_group',
       'assigned_to', 'notify', 'problem_id', 'rfc', 'vendor', 'caused_by',
       'closed_code', 'resolved_by', 'resolved_at', 'closed_at'],
      dtype='str')


C:\Users\aryan\AppData\Local\Temp\ipykernel_12364\1243307272.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cols=df.select_dtypes(include='object').columns


Converting ? to NAN VALUES

In [178]:
df.replace('?',pd.NA,inplace=True)

,number,incident_state,active,reassignment_count,reopen_count,sys_mod_count,made_sla,caller_id,opened_by,opened_at,...,u_priority_confirmation,notify,problem_id,rfc,vendor,caused_by,closed_code,resolved_by,resolved_at,closed_at
0,INC0000045,New,True,0,0,0,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,NaN,NaN,NaN,NaN,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
1,INC0000045,Resolved,True,0,0,2,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,NaN,NaN,NaN,NaN,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
2,INC0000045,Resolved,True,0,0,3,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,NaN,NaN,NaN,NaN,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
3,INC0000045,Closed,False,0,0,4,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,NaN,NaN,NaN,NaN,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
4,INC0000047,New,True,0,0,0,True,Caller 2403,Opened by 397,29/2/2016 04:40,...,False,Do Not Notify,NaN,NaN,NaN,NaN,code 5,Resolved by 81,1/3/2016 09:52,6/3/2016 10:00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
141707,INC0120835,Closed,False,1,0,4,True,Caller 116,Opened by 12,16/2/2017 09:09,...,True,Do Not Notify,NaN,NaN,NaN,NaN,code 9,Resolved by 9,16/2/2017 09:53,16/2/2017 09:53
141708,INC0121064,Active,True,0,0,0,True,Caller 116,Opened by 12,16/2/2017 14:17,...,False,Do Not Notify,NaN,NaN,NaN,NaN,code 6,Resolved by 9,16/2/2017 16:38,16/2/2017 16:38
141709,INC0121064,Active,True,1,0,1,True,Caller 116,Opened by 12,16/2/2017 14:17,...,False,Do Not Notify,NaN,NaN,NaN,NaN,code 6,Resolved by 9,16/2/2017 16:38,16/2/2017 16:38
141710,INC0121064,Resolved,True,1,0,2,True,Caller 116,Opened by 12,16/2/2017 14:17,...,True,Do Not Notify,NaN,NaN,NaN,NaN,code 6,Resolved by 9,16/2/2017 16:38,16/2/2017 16:38


In [179]:
df.isnull().sum()

number                          0
incident_state                  0
active                          0
reassignment_count              0
reopen_count                    0
sys_mod_count                   0
made_sla                        0
caller_id                      29
opened_by                    4835
opened_at                       0
sys_created_by              53076
sys_created_at              53076
sys_updated_by                  0
sys_updated_at                  0
contact_type                    0
location                       76
category                       78
subcategory                   111
u_symptom                   32964
cmdb_ci                    141267
impact                          0
urgency                         0
priority                        0
assignment_group            14213
assigned_to                 27496
knowledge                       0
u_priority_confirmation         0
notify                          0
problem_id                 139417
rfc           

Changing the date time values format to datetime

In [180]:
date_columns=['opened_at','sys_created_at','sys_updated_at','resolved_at']
for col in date_columns:
    df[col]=pd.to_datetime(df[col],errors='coerce')

C:\Users\aryan\AppData\Local\Temp\ipykernel_12364\2884995440.py:3: UserWarning: Parsing dates in %d/%m/%Y %H:%M format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df[col]=pd.to_datetime(df[col],errors='coerce')
C:\Users\aryan\AppData\Local\Temp\ipykernel_12364\2884995440.py:3: UserWarning: Parsing dates in %d/%m/%Y %H:%M format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df[col]=pd.to_datetime(df[col],errors='coerce')
C:\Users\aryan\AppData\Local\Temp\ipykernel_12364\2884995440.py:3: UserWarning: Parsing dates in %d/%m/%Y %H:%M format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df[col]=pd.to_datetime(df[col],errors='coerce')
C:\Users\aryan\AppData\Local\Temp\ipykernel_12364\2884995440.py:3: UserWarning: Parsing dates in %d/%m/%Y %H:%M format when dayfirst=False (the defa

In [181]:
df[date_columns].dtypes

opened_at         datetime64[us]
sys_created_at    datetime64[us]
sys_updated_at    datetime64[us]
resolved_at       datetime64[us]
dtype: object

In [182]:
print(df.duplicated().sum())

0


In [183]:
df.drop('cmdb_ci', axis=1, inplace=True, errors='ignore')

In [184]:
#deleting closed_at column
df.drop(columns=['closed_at'], inplace=True, errors='ignore')

In [185]:
#creating a temporary missing dataframe
missing=pd.DataFrame({'missing':df.isnull().sum(),
                      'missing_percentage':df.isnull().mean()*100})
missing=missing.sort_values(by='missing_percentage',ascending=False)
print(missing)

                         missing  missing_percentage
caused_by                 141689           99.983770
vendor                    141468           99.827820
rfc                       140721           99.300694
problem_id                139417           98.380518
sys_created_at             53076           37.453427
sys_created_by             53076           37.453427
u_symptom                  32964           23.261262
assigned_to                27496           19.402732
assignment_group           14213           10.029496
opened_by                   4835            3.411849
resolved_at                 3141            2.216467
closed_code                  714            0.503839
resolved_by                  226            0.159478
subcategory                  111            0.078328
category                      78            0.055041
location                      76            0.053630
caller_id                     29            0.020464
number                         0            0.

In [186]:
#analyzing first problematic column
df['caused_by'].head()

0    NaN
1    NaN
2    NaN
3    NaN
4    NaN
Name: caused_by, dtype: str

notna() helps us to tell exactly how many records have a caused_by value

In [187]:
df['caused_by'].notna().sum()

np.int64(23)

Checking exactly at which number the caused_by has value through label based selection i.e df.loc

In [188]:
df.loc[df['caused_by'].notna(),['number','caused_by']]

,number,caused_by
1174,INC0000388,CHG0000132
1175,INC0000388,CHG0000132
1176,INC0000388,CHG0000132
1177,INC0000388,CHG0000132
1178,INC0000388,CHG0000132
1179,INC0000388,CHG0000132
1180,INC0000388,CHG0000132
4512,INC0001131,CHG0000097
4513,INC0001131,CHG0000097
4514,INC0001131,CHG0000097


Dropping caused_by -since the feature is extremely sparse ,making it unsuitable as a reliable predictive feature for the model

In [189]:
df.drop(columns=['caused_by'],inplace=True)

Analyzing vendor column before dropping

In [190]:
df['vendor'].head(5)

0    NaN
1    NaN
2    NaN
3    NaN
4    NaN
Name: vendor, dtype: str

In [191]:
#printing non missing values
df['vendor'].notna().sum()

np.int64(244)

In [192]:
df.loc[df['vendor'].notna(),['number','vendor']]

,number,vendor
7857,INC0001845,code 8s
7858,INC0001845,code 8s
9086,INC0002099,code 8s
9087,INC0002099,code 8s
9088,INC0002099,code 8s
...,...,...
137209,INC0033686,Vendor 1
137210,INC0033686,Vendor 1
137211,INC0033686,Vendor 1
137212,INC0033686,Vendor 1


In [193]:
df['vendor'].nunique()

4

In [194]:
df['vendor'].value_counts()

vendor
code 8s     167
Vendor 1     69
Vendor 3      6
Vendor 2      2
Name: count, dtype: int64

In ITSM point of view a vendor is an external company/provider that supplies hardware,software,a product,service,or technical support to one organization

141,712 rows ≠ 141,712 incidents,as this is event log dataset,the same incident may appear multiple times.So we should check how many unique incidents have a vendor recorded.
 

In [195]:
df.loc[df['vendor'].notna(),'number'].unique()

<ArrowStringArray>
['INC0001845', 'INC0002099', 'INC0002104', 'INC0002332', 'INC0002336',
 'INC0002559', 'INC0003216', 'INC0003228', 'INC0003258', 'INC0004123',
 'INC0005008', 'INC0005213', 'INC0006165', 'INC0009186', 'INC0009537',
 'INC0010272', 'INC0010418', 'INC0011175', 'INC0011177', 'INC0011178',
 'INC0012044', 'INC0012984', 'INC0013219', 'INC0013636', 'INC0013833',
 'INC0013946', 'INC0015087', 'INC0015098', 'INC0015283', 'INC0015300',
 'INC0015847', 'INC0015927', 'INC0015945', 'INC0015951', 'INC0015953',
 'INC0016021', 'INC0016565', 'INC0016783', 'INC0016889', 'INC0016942',
 'INC0017106', 'INC0019197', 'INC0019739', 'INC0021267', 'INC0023726',
 'INC0023739', 'INC0024173', 'INC0024227', 'INC0024381', 'INC0024839',
 'INC0025213', 'INC0025982', 'INC0026235', 'INC0026474', 'INC0027949',
 'INC0028354', 'INC0028417', 'INC0029765', 'INC0029905', 'INC0029950',
 'INC0030004', 'INC0030649', 'INC0030708', 'INC0031444', 'INC0031532',
 'INC0032136', 'INC0033686']
Length: 67, dtype: str

Excluding vendor since vendor information is only available for 67 incidents out of 24918 unique incidents,indicating that vendor association is recorded only for very few incidents

In [196]:
df.drop(columns=['vendor'],inplace=True)

analyzing rfc feature before dropping

In [197]:
df['rfc'].nunique()

181

In [198]:
df['rfc'].notna().sum()

np.int64(991)

In [199]:
df.loc[df['rfc'].notna(),['number']].nunique()

number    179
dtype: int64

Excluding rfc since its having only 179 rfc information out of 24,918 unique incidents.Due to limited availability dropping the column

In [200]:
df.drop(columns=['rfc'],inplace=True)

Analyzing problem_id before dropping.problem_id tells which problem formally recorded problem is associated with the incident.

It is useful for root-cause analysis and problem management.

In [201]:
df['problem_id'].nunique()

252

In [202]:
df['problem_id'].notna().sum()

np.int64(2295)

In [203]:
df.loc[df['problem_id'].notna(),'number'].nunique()

381

Dropped problem_id because approximately 98.38% of its values were missing. Problem information was available for only 381 out of 24,918 unique incidents (~1.53%)

In [204]:
df.drop(
    columns=['caused_by', 'vendor', 'rfc', 'problem_id'],
    inplace=True,
    errors='ignore'
)

In [205]:
missing = df.isnull().sum()

missing_percentage = (missing / len(df)) * 100

missing_df = pd.DataFrame({
    'missing': missing,
    'missing_percentage': missing_percentage
})

print(missing_df.sort_values(by='missing_percentage', ascending=False))



                         missing  missing_percentage
sys_created_by             53076           37.453427
sys_created_at             53076           37.453427
u_symptom                  32964           23.261262
assigned_to                27496           19.402732
assignment_group           14213           10.029496
opened_by                   4835            3.411849
resolved_at                 3141            2.216467
closed_code                  714            0.503839
resolved_by                  226            0.159478
subcategory                  111            0.078328
category                      78            0.055041
location                      76            0.053630
caller_id                     29            0.020464
number                         0            0.000000
sys_updated_at                 0            0.000000
sys_updated_by                 0            0.000000
made_sla                       0            0.000000
opened_at                      0            0.

In [206]:
# Drop columns with more than 95% missing values

df.drop(
    columns=['caused_by', 'vendor', 'rfc', 'problem_id'],
    inplace=True,
    errors='ignore'
)

print("Columns dropped successfully (if they existed).")

Columns dropped successfully (if they existed).


In [207]:
df.drop(
    columns=['caused_by', 'vendor', 'rfc', 'problem_id'],
    inplace=True,
    errors='ignore'
)

In [208]:
missing = df.isnull().sum()

missing_percentage = (missing / len(df)) * 100

missing_df = pd.DataFrame({
    "Missing": missing,
    "Missing %": missing_percentage
})

missing_df = missing_df[missing_df["Missing"] > 0]
missing_df = missing_df.sort_values("Missing %", ascending=False)

missing_df

,Missing,Missing %
sys_created_by,53076,37.453427
sys_created_at,53076,37.453427
u_symptom,32964,23.261262
assigned_to,27496,19.402732
assignment_group,14213,10.029496
opened_by,4835,3.411849
resolved_at,3141,2.216467
closed_code,714,0.503839
resolved_by,226,0.159478
subcategory,111,0.078328


In [209]:
print(df.columns.tolist())

['number', 'incident_state', 'active', 'reassignment_count', 'reopen_count', 'sys_mod_count', 'made_sla', 'caller_id', 'opened_by', 'opened_at', 'sys_created_by', 'sys_created_at', 'sys_updated_by', 'sys_updated_at', 'contact_type', 'location', 'category', 'subcategory', 'u_symptom', 'impact', 'urgency', 'priority', 'assignment_group', 'assigned_to', 'knowledge', 'u_priority_confirmation', 'notify', 'closed_code', 'resolved_by', 'resolved_at']


In [210]:
df.drop(
    columns=['caused_by', 'vendor', 'rfc', 'problem_id'],
    errors='ignore',
    inplace=True
)

In [211]:
print(df.shape)

(141712, 30)


In [212]:
# Fill categorical columns with meaningful values

df["assigned_to"] = df["assigned_to"].fillna("Unassigned")
df["assignment_group"] = df["assignment_group"].fillna("Unassigned")

df["opened_by"] = df["opened_by"].fillna("Unknown")
df["resolved_by"] = df["resolved_by"].fillna("Unknown")
df["location"] = df["location"].fillna("Unknown")
df["caller_id"] = df["caller_id"].fillna("Unknown")

# Fill with the most frequent value (mode)
df["category"] = df["category"].fillna(df["category"].mode()[0])
df["subcategory"] = df["subcategory"].fillna(df["subcategory"].mode()[0])

# Symptom column
df["u_symptom"] = df["u_symptom"].fillna("Unknown")

In [213]:
# Convert to datetime
df["resolved_at"] = pd.to_datetime(df["resolved_at"], errors="coerce")

# Fill missing dates with the most frequent date
df["resolved_at"] = df["resolved_at"].fillna(df["resolved_at"].mode()[0])

In [214]:
print(df["sys_created_by"].dtype)
print(df["sys_created_at"].dtype)

str
datetime64[us]


In [215]:
df["sys_created_by"] = df["sys_created_by"].fillna("Unknown")
df["sys_created_at"] = df["sys_created_at"].fillna("Unknown")

In [216]:
missing = df.isnull().sum()

missing = missing[missing > 0]

print(missing)

closed_code    714
dtype: int64


In [217]:
print(df[["sys_created_by", "sys_created_at"]].dtypes)

sys_created_by       str
sys_created_at    object
dtype: object


In [218]:
missing = df.isnull().sum()
missing = missing[missing > 0]
print(missing)

closed_code    714
dtype: int64


In [219]:
print(df["closed_code"].dtype)


str


In [220]:
df["closed_code"] = df["closed_code"].fillna(df["closed_code"].mode()[0])

In [221]:
df["closed_code"] = df["closed_code"].fillna("Unknown")

Verify all missing values are handled

In [222]:
missing = df.isnull().sum()
missing = missing[missing > 0]
print(missing)


Series([], dtype: int64)


missing value treatment is complete.

Now Check and remove duplicate rows.

In [223]:
# Check duplicate rows
duplicate_count = df.duplicated().sum()

print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 0


What this does:
df.duplicated() checks whether each row is identical to a previous row.
.sum() counts how many duplicate rows exist.

In [224]:
# Remove duplicate rows
df.drop_duplicates(inplace=True)

print("Duplicate rows removed successfully.")

Duplicate rows removed successfully.


Remove duplicates

In [225]:
print("Number of duplicate rows:", df.duplicated().sum())

Number of duplicate rows: 0


Dataset does not contain any duplicate rows

In [226]:
print(df.select_dtypes(include=['datetime64[ns]']).columns)

Index(['opened_at', 'sys_updated_at', 'resolved_at'], dtype='str')


Convert to datetime

In [227]:
date_columns = ['opened_at', 'sys_updated_at', 'resolved_at']

for col in date_columns:
    df[col] = pd.to_datetime(df[col], errors='coerce')

print(df[date_columns].dtypes)

opened_at         datetime64[us]
sys_updated_at    datetime64[us]
resolved_at       datetime64[us]
dtype: object


Create new date features

In [228]:
for col in date_columns:
    df[col + '_year'] = df[col].dt.year
    df[col + '_month'] = df[col].dt.month
    df[col + '_day'] = df[col].dt.day
    df[col + '_hour'] = df[col].dt.hour
    df[col + '_weekday'] = df[col].dt.day_name()

Verify the new columns

In [229]:
print(df.head())

       number incident_state  active  reassignment_count  reopen_count  \
0  INC0000045            New    True                   0             0   
1  INC0000045       Resolved    True                   0             0   
2  INC0000045       Resolved    True                   0             0   
3  INC0000045         Closed   False                   0             0   
4  INC0000047            New    True                   0             0   

   sys_mod_count  made_sla    caller_id       opened_by           opened_at  \
0              0      True  Caller 2403    Opened by  8 2016-02-29 01:16:00   
1              2      True  Caller 2403    Opened by  8 2016-02-29 01:16:00   
2              3      True  Caller 2403    Opened by  8 2016-02-29 01:16:00   
3              4      True  Caller 2403    Opened by  8 2016-02-29 01:16:00   
4              0      True  Caller 2403  Opened by  397 2016-02-29 04:40:00   

   ... sys_updated_at_year sys_updated_at_month sys_updated_at_day  \
0  ...    

now Encoding Categorical Variables.

Find all categorical columns

In [230]:
cat_cols = df.select_dtypes(include='object').columns

print(cat_cols)
print("Total categorical columns:", len(cat_cols))

Index(['number', 'incident_state', 'caller_id', 'opened_by', 'sys_created_by',
       'sys_created_at', 'sys_updated_by', 'contact_type', 'location',
       'category', 'subcategory', 'u_symptom', 'impact', 'urgency', 'priority',
       'assignment_group', 'assigned_to', 'notify', 'closed_code',
       'resolved_by', 'opened_at_weekday', 'sys_updated_at_weekday',
       'resolved_at_weekday'],
      dtype='str')
Total categorical columns: 23


C:\Users\aryan\AppData\Local\Temp\ipykernel_12364\1799255628.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cat_cols = df.select_dtypes(include='object').columns


Choose the encoding method